core/src/main/java/io/grpc/internal/MessageDeframer.java

```diff
@@ -438,6 +449,18 @@ public synchronized void reset() throws IOException {
       count = mark;
     }
 
+    @Override
+    public void close() throws IOException {
+      super.close();
+    }
+
+    private void reportCount() {
+      if (count > maxCount) {
+        statsTraceContext.messageReceived(count - maxCount);
```

Comment:
In finagle tracing, there's a wire fragment received annotation that is
logged as a trace annotation once for each chunk of a single message (in
the same span).

If message received acts like this, at least in tools like zipkin, it can
be visualised.

If a "message" here is more semantic in nature, ex new operation, I suspect
it might imply provisioning new span ids and could be more problematic or
at least complicated.

On 14 Sep 2016 11:39 pm, "Kun Zhang" <notifications@github.com> wrote:

> In core/src/main/java/io/grpc/internal/MessageDeframer.java
> <https://github.com/grpc/grpc-java/pull/2262#discussion_r78774932>:
>
> > @@ -438,6 +449,18 @@ public synchronized void reset() throws IOException {
> >        count = mark;
> >      }
> >
> > +    @Override
> > +    public void close() throws IOException {
> > +      super.close();
> > +    }
> > +
> > +    private void reportCount() {
> > +      if (count > maxCount) {
> > +        statsTraceContext.messageReceived(count - maxCount);
>
> This is problematic. I wanted to have messageReceived() called once for
> each message, so that if tracing will want to expose message reading
> events, it will get it right. Now it may be recorded as multiple
> messageReceived events, depending on how fine-grained it is read.
>
> Alternatively, I could do this in close(), but that requires all
> Marshallers to close the InputStream. No Marshaller does it today. This
> would be an API change. Not that it would break users, but that stats won't
> work correctly if the implementation doesn't follow the new API. I think it
> should be fine. @ejona86 <https://github.com/ejona86> WDYT?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc-java/pull/2262/files/5edccda621103fde718491feb4bca6c85a4f24ac#r78774932>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD61xXuZWuQsmOD8wP9i1ZF_i2Jv0Uvks5qqBVGgaJpZM4J75sK>
> .
>